Feature 159: Workgroup Berechtigungen

daniel-va opened the pull request:
Resolves #159.

Rewrite of the application's authorization structure, which is now based on the workgroups assigned to the current user.

- Authorization is backed by `Policy` classes, which define if a user is authorized for an action on a specific entity.
- To configure authorization, the new `Authorize` annotations are now attached to controller routes.
- The new models previously found in the backend have been moved into the new `@shared` library, which can be found at `libs/shared2`. See its README for more details on the library's purpose.
- Removal of the `User.role` field, which is now no longer used.
- In the frontend, the new directives `adminOnly`, `canShow`, `canCreate`, `canUpdate` and `canDelete` can be used to conditionally render HTML. They are backed by the same policies used on the backend.
- The PR includes deletions of unused old code which is no longer in use.

TIL-EBP commented on apps/server-asset-sg/src/features/users/user.model.ts:
```diff
@@ -1,37 +1,23 @@
 export interface WorkgroupOnUser {
-  workgroupId: number;
-  role: PrismaRole;
-  workgroup: {
```
shouldnt we keep the name for easier handling in the frontend?

daniel-va replied:
I'm currently expecting that we will load all available workgroups via reference-data, which would mean that we don't need the name in here, as we can get the whole workgroup via its id.

TIL-EBP commented on apps/server-asset-sg/src/features/favorites/favorites.controller.ts:
```diff
@@ -23,12 +25,14 @@ export class FavoritesController {
   }
 
   @Put('/:assetId')
```
So users can potentially add assets to theri favourits which they are not allowed to see because of lacking workgroup access? Would this be a loophole to view assets if (of course only if i know the id despite not being allowed to view it)

daniel-va replied:
Favorites are not yet correctly authorized as they are not really implemented. If we ever get to them, we would probably write a new policy just for them.

TIL-EBP commented on apps/server-asset-sg/src/features/contacts/contacts.controller.ts:
```diff
@@ -10,18 +10,16 @@ import {
 
   @Post('/')
-  @RequireRole(Role.Editor)
+  @Authorize.Admin()
```
why can only admin add new contacts? sholdn't editors also be allowed to do this?

TIL-EBP commented on apps/server-asset-sg/src/features/assets/asset.policy.ts:
```diff
@@ -0,0 +1,25 @@
+
+  canCreate(): boolean {
+    // A user can create assets for workgroups for which they are an Editor.
+    return null != this.user.workgroups.find((group) => getRoleIndex(group.role) >= getRoleIndex(Role.Editor));
```
Dont we need to check to which workgroup the user wants to add the asset? This simply checks whether a user is editor in any of their workgroups right?

daniel-va replied:
Yes, this is done in the `AssetEditController.create` method. The policy does not have access to the request body, which is why we can only check if the user has the ability to create assets in general. This has the downside of requiring an extra validation of the request body. On the other hand, due to this, we can reuse the policy in other places were there is no body (e.g. in the frontend form, or in other controllers).

daniel-va commented on apps/server-asset-sg/src/core/policy.ts:
```diff
@@ -0,0 +1,56 @@
+    return this.canCreate();
+  }
+
+  canDelete(value: T): boolean {
```
Underscore means that the parameter is ignored, i.e. it does not count as unused even if it is. This is useful in situations were you want a parameter (for documentation and/or because you might need it later on) but don't want to use it.

daniel-va commented on apps/server-asset-sg/src/features/workgroups/workgroup.controller.ts:
```diff
@@ -1,72 +1,61 @@
-    }
-    return workGroup;
+  @Get('/')
+  @Authorize.User()
```
Yes, this just means that all users can access the API path that reads the workgroups. The list is then filtered within the route so that the user only gets the assets that he has access to.

TIL-EBP replied:
Yeah I get that, but this still means that every user can get a list of all the Workgroups, right? Maybe they should not be able to see which wokrgroups exist (even though this information does not get them anywhere)

daniel-va replied:
@TIL-EBP is this cleared up or would you prefer to talk about this in more detail?

daniel-va commented on apps/server-asset-sg/src/features/asset-edit/asset-edit.policy.ts:
```diff
@@ -0,0 +1,25 @@
+import { AssetEditDetail } from '@asset-sg/shared';
```
Yes, its exactly the same. Currently, there are two versions of assets - `AssetEditDetail` and the newer `Asset`, which is actually not used anywhere in the frontend (yet?).
The main difference, beside their different structure, is that `AssetEditDetail` is based on fp-ts, while `Asset` is not.

vej-ananas commented:
From manual testing:
- On inital page load, it seems like the User is not requested and the behavior with e.g. the menus on the left is inconsistent
- When I try to access an asset I am not authorized to (Im not Admin, Asset in other Workgroup), I get thrown out of the application. I would expect a 404 screen or similar
![image](https://github.com/user-attachments/assets/c0a8b95c-3bea-4b2c-9691-bdc90eb47bf0)

- When I try to access the Edit page (e.g. http://localhost:4200/de/asset-admin/44374/general), I am stuck on a white screen. I would expect 404 or Home screen
- I was unable to edit the Status as Master-Editor with the error message `Changing the asset's internal status is not allowed`

I did not see any unit tests - did you consider writing any?

daniel-va commented:
@vej-ananas 

> From manual testing:
> 
>     * On inital page load, it seems like the User is not requested and the behavior with e.g. the menus on the left is inconsistent

Fixed.

 
>     * When I try to access an asset I am not authorized to (Im not Admin, Asset in other Workgroup), I get thrown out of the application. I would expect a 404 screen or similar
>       ![image](https://private-user-images.githubusercontent.com/103118462/350007328-c0a8b95c-3bea-4b2c-9691-bdc90eb47bf0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNjc5ODcsIm5iZiI6MTcyMTM2NzY4NywicGF0aCI6Ii8xMDMxMTg0NjIvMzUwMDA3MzI4LWMwYThiOTVjLTNiZWEtNGIyYy05NjkxLWJkYzkwZWI0N2JmMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQwNTQxMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMjc3MDNjYjZmMWJlOTNiNzZjNWVmZmJiMjE2ZjA3ODFiNmRlOGMwYmMyMThmNWZiZTU1Y2MwM2JkZTI5M2U2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.cQc8_Ax5KgY-kKxSLRfgvILa8IJCBrFXcqSVOhLZseA)

This behaviour is synonymous to the current live version. If we change this, we should probably define one shared way missing permissions are handled everywhere in the app.

>     * When I try to access the Edit page (e.g. http://localhost:4200/de/asset-admin/44374/general), I am stuck on a white screen. I would expect 404 or Home screen

Fixed.

>     * I was unable to edit the Status as Master-Editor with the error message `Changing the asset's internal status is not allowed`

Fixed.

> I did not see any unit tests - did you consider writing any?

TBD.

daniel-va commented on libs/client-shared/src/lib/directives/adminOnly.directive.ts:
The [Angular style guide](https://v17.angular.io/guide/styleguide#directive-selectors) recommends lower camel case for directive selectors.

vej-ananas replied:
sure but not in the file name - we always have the file name kebab even if the class or directive name follows a different convention

daniel-va commented on apps/client-asset-sg/src/environments/environment.ts:
Ich lass es mal so wenn's für dich auch passt, falls es jemand braucht kann man es ja selbst wieder aktivieren.

daniel-va commented on apps/server-asset-sg/src/features/asset-edit/asset-edit.controller.ts:
```diff
@@ -0,0 +1,85 @@
+   */
+  @Post('/')
+  @Authorize.Create()
+  async create(
```
This route can't currently removed, the deprecation notice is actually wrong. I've removed it.

daniel-va commented on apps/server-asset-sg/src/features/assets/search/asset-search.service.ts:
```diff
@@ -161,17 +162,19 @@ export class AssetSearchService {
    */
   async search(
     query: AssetSearchQuery,
+    user: User,
```
I would leave it like this for now and go over it again in #163. It might even be possible to fully move the user/its workgroups into the query, so that we don't need the extra parameter.

daniel-va commented on libs/shared2/README.md:
We can, but then have an overlap of the `models` directories. `lib/shared` contains quite a bit of outdated code, and some parts of it are duplicated among other libraries. I'm expecting that at some point, most parts of it will be obsolete/replaced, which is when we can fully remove it.

With `lib/shared2`, I wanted to indicate that it is a clean collection of shared, up-to-date code.

vej-ananas commented:
> @vej-ananas
> 
> > From manual testing:
> > ```
> > * On inital page load, it seems like the User is not requested and the behavior with e.g. the menus on the left is inconsistent
> > ```
> 
> Fixed.

Initial Load looks good - after some reloads via F5 when I am not Admin, the `Verwaltung` button disappears.

> 
> > ```
> > * When I try to access an asset I am not authorized to (Im not Admin, Asset in other Workgroup), I get thrown out of the application. I would expect a 404 screen or similar
> >   ![image](https://private-user-images.githubusercontent.com/103118462/350007328-c0a8b95c-3bea-4b2c-9691-bdc90eb47bf0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNjc5ODcsIm5iZiI6MTcyMTM2NzY4NywicGF0aCI6Ii8xMDMxMTg0NjIvMzUwMDA3MzI4LWMwYThiOTVjLTNiZWEtNGIyYy05NjkxLWJkYzkwZWI0N2JmMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQwNTQxMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMjc3MDNjYjZmMWJlOTNiNzZjNWVmZmJiMjE2ZjA3ODFiNmRlOGMwYmMyMThmNWZiZTU1Y2MwM2JkZTI5M2U2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.cQc8_Ax5KgY-kKxSLRfgvILa8IJCBrFXcqSVOhLZseA)
> > ```
> 
> This behaviour is synonymous to the current live version. If we change this, we should probably define one shared way missing permissions are handled everywhere in the app.
> 

Lets do that! Considering our users are authenticated in our app I do not find it critical to prevent knowledge of an existing id - so rather 403 than 404, right?

> > ```
> > * When I try to access the Edit page (e.g. http://localhost:4200/de/asset-admin/44374/general), I am stuck on a white screen. I would expect 404 or Home screen
> > ```
> 
> Fixed.

Similarly to above: fixed as initial url, but when already on the page it fails.

> 
> > ```
> > * I was unable to edit the Status as Master-Editor with the error message `Changing the asset's internal status is not allowed`
> > ```
> 
> Fixed.
> 
> > I did not see any unit tests - did you consider writing any?
> 
> TBD.

vej-ananas commented on libs/auth/src/lib/services/auth.interceptor.ts:
```diff
@@ -41,7 +41,9 @@ export class AuthInterceptor implements HttpInterceptor {
       case 403:
-        this.authService.setState(AuthState.Forbidden);
+        this.authService.setState(
+          error.error == 'not authorized by eIAM' ? AuthState.AccessForbidden : AuthState.ForbiddenResource
```
I dont think ForbiddenResource should be an application state - it is not the whole application but a single resource.
Currently, once you get this error, your only way out is to reload the whole application.

vej-ananas commented on apps/client-asset-sg/project.json:
```diff
@@ -19,7 +19,7 @@
         "styles": ["apps/client-asset-sg/src/styles.scss"],
         "scripts": [],
-        "allowedCommonJsDependencies": ["tsafe", "validator", "xml-utils", "pbf", "rbush", "earcut"]
+        "allowedCommonJsDependencies": ["tsafe", "validator", "xml-utils", "pbf", "rbush", "earcut", "@prisma/client"]
```
I imagine this was unintentional?
